include/wil/registry_basic.h

```diff
@@ -0,0 +1,1540 @@
+
+            DWORD disposition = 0;
+            const auto error =
+                ::RegCreateKeyExW(::wil::registry::details::get_key(m_key), subKey, 0, nullptr, 0, ::wil::registry::details::get_access_flags(access), security_descriptor ? &security_attributes : nullptr, ::wil::out_param(hkey), &disposition);
+            return err_policy::HResult(HRESULT_FROM_WIN32(error));
+        }
+
+#if defined(__WIL_WINREG_STL)
+    typename err_policy::result create_key(_In_ PCWSTR subKey, ::wil::shared_hkey& hkey, ::wil::registry::key_access access = ::wil::registry::key_access::read, _In_opt_ PCWSTR security_descriptor = nullptr) const
+    {
+        hkey.reset();
+
+        const ::wil::unique_hlocal_ptr<SECURITY_DESCRIPTOR> securityDescriptor = ::wil::registry::details::create_security_descriptor<err_policy>(security_descriptor);
+
+        SECURITY_ATTRIBUTES security_attributes{sizeof security_attributes, nullptr, FALSE};
+        security_attributes.lpSecurityDescriptor = securityDescriptor.get();
+
+        DWORD disposition = 0;
+        const auto error =
+            ::RegCreateKeyExW(::wil::registry::details::get_key(m_key), subKey, 0, nullptr, 0, ::wil::registry::details::get_access_flags(access), security_descriptor ? &security_attributes : nullptr, ::wil::out_param(hkey), &disposition);
+        return err_policy::HResult(HRESULT_FROM_WIN32(error));
+    }
+#endif
+
+        ::wil::unique_hkey create_unique_key(_In_ PCWSTR subKey, ::wil::registry::key_access access = ::wil::registry::key_access::read, _In_opt_ PCWSTR security_descriptor = nullptr) const
+        {
+            ::wil::unique_hkey local_key{};
+            create_key(subKey, local_key, access, security_descriptor);
+            return local_key;
+        }
+
+#if defined(__WIL_WINREG_STL)
+    ::wil::shared_hkey create_shared_key(_In_ PCWSTR subKey, ::wil::registry::key_access access = ::wil::registry::key_access::read, _In_opt_ PCWSTR security_descriptor = nullptr) const
+    {
+        ::wil::shared_hkey local_key{};
+        create_key(subKey, local_key, access, security_descriptor);
+        return local_key;
+    }
+#endif
+
+        typename err_policy::result delete_key(_In_opt_ PCWSTR sub_key) WI_NOEXCEPT
+        {
+            auto error = RegDeleteTreeW(::wil::registry::details::get_key(m_key), sub_key);
+            if (error == ERROR_FILE_NOT_FOUND)
+            {
+                error = ERROR_SUCCESS;
+            }
+            return err_policy::HResult(HRESULT_FROM_WIN32(error));
+        }
+
+        typename err_policy::result delete_value(HKEY key, _In_opt_ PCWSTR value_name) WI_NOEXCEPT
+        {
+            return err_policy::HResult(HRESULT_FROM_WIN32(RegDeleteValueW(key, value_name)));
+        }
+
+        template <typename R>
+        typename err_policy::result get_value(_In_opt_ PCWSTR subkey, _In_opt_ PCWSTR value_name, optional_value<R>& return_value, DWORD type = ::wil::registry::reg_view_details::get_value_type<R>()) const

```

Comment:
We discussed this, but I want to separate this into `get_value` and `try_get_value`, where `get_value` errors if the value is missing.

That way callers that expect a value to exist can fail quicker.

Reply:
we should probably discuss this more - it looks like wil tend to return a bool (or something convertable to bool) from its try_* functions. If that's the case, we'll have to rework these.

Looking through this again, I think I want to make this more consistent so that get* calls either return:
- HRESULT for no-throw, and they pass an OUT ptr for the type they want to read (if they really want to deal with all errors, including value-not-found).
- bool for try_get (or maybe return optional? since that can be convertable to bool?)
- not sure what to return now for just get_value() that doesn't return an HRESULT if try_get_value is returning an optional. I don't know that I want to have the 'default' get_value throw when the value isn't present.

Reply:
My hunch is that the prior art in wil was for lack of a good `optional` type. It's pretty idiomatic to return an `optional<T>` in modern C++, plus it avoids the outptr shenanigans.

I think the approach we've ended up with is good:

* `*_nothrow` returns `HRESULT` (including not found), uses outptr.
* `try_get_*` returns `optional<T>` (empty when not found; throws on other errors), no outptr.
* `get_*` returns `T` (throws on all errors, including not found), no outptr.

This makes sense: if a caller expects a value to exist, it is an error for it not to exist.

Reply:
(Since we talked about it: the Rusty way would be to always return an optional and have callers `.expect()` it 😏 )